deepform/data/add_features.py

```diff
@@ -132,30 +134,31 @@ def process_document_tokens(token_file, dest_file, labels):
     # Return the summary information about the document.
     return {"slug": slug, "length": len(doc), **labels, **best_matches}
 
-## This is the function as it currently exists:
-def label_tokens(tokens, labels, n):
+
+def label_tokens(tokens, labels, max_token_count):
     for col_name, label_value in labels.items():
         tokens[col_name] = 0.0
         match_fn = LABEL_COLS[col_name]
 
-    # Assemble all the token strings that can be formed for a token at a specific index with maximum n-gram length of n
-        for index, token in tokens.token.items():  
+        # Assemble all the token strings that can be formed for a token
+        # at a specific index with maximum n-gram length of max_token_count
+        for i in range(len(tokens)):
             n_grams = []
             match_percentages = []
-            for x in range(1,n+1): # x is all the possible lengths of n-gram
-                for y in range (1, x+1): # y is all the possible index modifications for that length of n-gram
-                    n_gram = tokens.loc[max(0,index-y+1):max(0,index-y+x), "token"].values 
-                    n_gram = ''.join(n_gram)
+            for length in range(1, max_token_count + 1):
+                for start in range(max(0, i - length + 1), i + 1):
+                    n_gram = tokens.token[start : start + length].values
+                    n_gram = " ".join(n_gram)
```

Comment:
You added this space here so in case of multi-token dates or $ amounts?  I believe the lhs and rhs of defaultsimilarity() both have simplestring() applied.  I agree with this since it shouldn't matter in the case of default and may help in the case of $ or date similarity.